src/hugr.rs

```diff
@@ -73,6 +77,127 @@ impl Hugr {
+                    for connection in r.n.graph.get_connections(*n_index, n_node_succ.index) {
+                        let src_offset = r.n.graph.port_offset(connection.0).unwrap().index();
+                        let tgt_offset = r.n.graph.port_offset(connection.1).unwrap().index();
+                        self.graph
+                            .link_nodes(
+                                *new_node_index,
+                                src_offset,
+                                *new_node_succ_index,
+                                tgt_offset,
+                            )
+                            .ok();
+                    }
+                }
+            }
+        }
+        // 3.3. For each p in inp(n_dfg_node), add an edge from the predecessor of r.nu_inp[p] to (new copy of) p.
+        for ((n_inp_node, n_inp_port), (s_inp_node, s_inp_port)) in r.nu_inp {
+            let new_inp_node_index = index_map.get(&n_inp_node.index).unwrap();
+            // add edge from predecessor of (s_inp_node, s_inp_port) to (new_inp_node, n_inp_port)
+            let s_inp_portindex = self
+                .graph
+                .port_index(s_inp_node.index, s_inp_port.offset)
+                .unwrap();
+            let s_preinp_portindex = self.graph.port_link(s_inp_portindex).unwrap();
+            let new_inp_portindex = self
+                .graph
+                .port_index(*new_inp_node_index, n_inp_port.offset)
+                .unwrap();
+            self.graph.unlink_port(s_preinp_portindex);
+            self.graph
+                .link_ports(s_preinp_portindex, new_inp_portindex)
+                .ok();
+        }
+        // 3.4. For each p in out(r.s), add an edge from (new copy of) the predecessor of r.nu_out[p] to p.
+        for ((s_exit_node, s_exit_port), n_out_port) in r.nu_out {
+            let s_exit_portindex = self
+                .graph
+                .port_index(s_exit_node.index, s_exit_port.offset)
+                .unwrap();
+            let n_out_portindex =
+                r.n.graph
+                    .port_index(n_output_node_index, n_out_port.offset)
+                    .unwrap();
+            let n_preexit_portindex = r.n.graph.port_link(n_out_portindex).unwrap();
+            let n_preexit_nodeindex = r.n.graph.port_node(n_preexit_portindex).unwrap();
+            let n_preexit_portoffset = r.n.graph.port_offset(n_preexit_portindex).unwrap();
+            let new_out_nodeindex = index_map.get(&n_preexit_nodeindex).unwrap();
+            let new_out_portindex = self
+                .graph
+                .port_index(*new_out_nodeindex, n_preexit_portoffset)
+                .unwrap();
+            self.graph.unlink_port(s_exit_portindex);
+            self.graph
+                .link_ports(new_out_portindex, s_exit_portindex)
+                .ok();
+        }
+        // 3.5. Remove all nodes in r.s and edges between them.
```

Comment:
Can we do this before inserting the new nodes, to allow reusing memory?

Reply:
No, because we then panic at the line
```
let rem_inp_predecessor_port_index = self.graph.port_link(rem_inp_port_index).unwrap();
```
where we need to locate the predecessor node.